segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.metadata;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.segmentstore.contracts.StreamSegmentExistsException;
+import io.pravega.segmentstore.contracts.tables.BadKeyVersionException;
+import io.pravega.segmentstore.contracts.tables.TableEntry;
+import io.pravega.segmentstore.contracts.tables.TableKey;
+import io.pravega.segmentstore.contracts.tables.TableStore;
+import io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * {@link TableStore} based storage metadata store.
+ */
+@Slf4j
+public class TableBasedMetadataStore extends BaseMetadataStore {
+    private final TableStore tableStore;
+    private final String tableName;
+    private final Duration timeout = Duration.ofSeconds(1L);
+    private final AtomicBoolean isTableInitialized = new AtomicBoolean(false);
+    private final BaseMetadataStore.TransactionData.TransactionDataSerializer serializer = new BaseMetadataStore.TransactionData.TransactionDataSerializer();
+
+    /**
+     * Constructor.
+     *
+     * @param tableName Name of the table segment.
+     * @param tableStore Instance of the {@link TableStore}.
+     */
+    public TableBasedMetadataStore(String tableName, TableStore tableStore) {
+        this.tableStore = Preconditions.checkNotNull(tableStore, "tableStore");
+        this.tableName = Preconditions.checkNotNull(tableName, "tableName");
+    }
+
+    /**
+     * Reads a metadata record for the given key.
+     *
+     * @param key Key for the metadata record.
+     * @return Associated {@link io.pravega.segmentstore.storage.metadata.BaseMetadataStore.TransactionData}.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    protected TransactionData read(String key) throws StorageMetadataException {
+        ensureInitialized();
+        List<BufferView> keys = new ArrayList<>();
+        keys.add(new ByteArraySegment(key.getBytes()));
+        try {
+            List<TableEntry> retValue = this.tableStore.get(tableName, keys, timeout).get();
+            if (retValue.size() == 1) {
```

Comment:
So what if it's not 1?

Reply:
The code falls through and it returns default "not exists" value.

Reply:
Ok. will add validation.

Reply:
Validation added.

Reply:
The contract clearly says that you will get a List of size equal to the size of the key List you requested, regardless of whether those keys exist or not. You will not get 0 and you will not get 2. 

If you want, a simple `assert retValue.size() == 1` will do (and will execute only in tests), and we can remove all these unnecessary checks.

Reply:
Fixed.  it looks like this now
 ``` java
            Preconditions.checkState(retValue.size() == 1, "Unexpected number of values returned.");
            TableEntry entry = retValue.get(0);
            if (null != entry) {
              ...
```